### PR TITLE
Make space & chat selectors searchable, fix several issues along the way

### LINKED ISSUE #1595: Linking space chat fails with "invalid Matrix content URI"
Trying to link an existing chat to a new space:
1. first loads for a very long time
2. then fails with the error message as shown below:

grafik

### LINKED ISSUE #1625: Implement search in space & chat selector widgets.

### PR DESCRIPTION
**Pull request by gnunicorn**
Massively refactors the chat & space selectors: clean up several providers that aren't useful in their current state as they fetch too much information not locally needed, fixes #1595 . While on it refactors the space selection drawer to include a similar search (if there are more than 10 items in the list) and make it a single widget used for both chat and space selection. Use the same Brief room view for all of these instances removes further code.

https://github.com/acterglobal/a3/assets/40496/25fa7e13-6b47-4a7e-89a3-1b2b36843f95

Fixes #1595 , fixes #1625 .

### CODE REVIEW

File: app/lib/common/widgets/room/select_room_drawer.dart [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+  groupChat,
+  space,
+}
+
+class SelectRoomDrawer extends ConsumerStatefulWidget {
+  final String canCheck;
+  final String? currentSpaceId;
+  final Widget title;
+  final String keyPrefix;
+  final DisplayMode avatarDisplayMode;
```
Comment (gtalha07) [inline app/lib/common/widgets/room/select_room_drawer.dart:24]: I feel like this param can be determined using ```roomType``` internally and doesn't necessarily need to be passed down...

File: app/lib/common/widgets/room/select_room_drawer.dart [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+import 'package:atlas_icons/atlas_icons.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:path/path.dart';
+import 'package:flutter_gen/gen_l10n/l10n.dart';
+
+// ChildRoomType configures the sub child type of the `Spaces`
+enum RoomType {
+  groupChat,
+  space,
```
Comment (gtalha07) [inline app/lib/common/widgets/room/select_room_drawer.dart:16]: we can add ```privateChat/DM``` value here too in reference to below refactor suggestion.

File: app/lib/common/widgets/room/select_room_drawer.dart (hunk 1 above)
Comment (gtalha07) [inline app/lib/common/widgets/room/select_room_drawer.dart:24]: I feel like this param can be determined internally and doesn't need to necessarily be passed down here...

File: app/lib/common/widgets/room/select_room_drawer.dart [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+import 'package:acter/common/widgets/room/brief_room_list_entry.dart';
+import 'package:acter_avatar/acter_avatar.dart';
+import 'package:atlas_icons/atlas_icons.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:path/path.dart';
+import 'package:flutter_gen/gen_l10n/l10n.dart';
+
+// ChildRoomType configures the sub child type of the `Spaces`
+enum RoomType {
```
Comment (gtalha07) [inline app/lib/common/widgets/room/select_room_drawer.dart:14]: This can determine avatar type based on below comment ;)

**Review by gtalha07 — APPROVED** [review]
Nice improvements on providers code! LGTM!!

**Review by kumarpalsinh25 — APPROVED** [review]
Nice Improvement!
Simplified many things! Changes looks good to me. :)

File: app/lib/common/widgets/room/brief_room_list_entry.dart [hunk 4]
```diff
@@ -9,16 +9,20 @@ class BriefRoomEntry extends ConsumerWidget {
   final String? selectedValue;
   final String canCheck;
   final String keyPrefix;
-  final Function(String) onSelect;
+  final Function(String)? onSelect;
   final DisplayMode avatarDisplayMode;
+  final Widget Function(bool)? trailingBuilder;
```
[the remark this answers is not in the document]
Comment (gnunicorn) [reply app/lib/common/widgets/room/brief_room_list_entry.dart:14]: they serve slightly different functions. the `canCheck` would enable the entire object and the trailingBuilder only provides a specific feature (the link and unlink). We could allow not providing a `canCheck` but in an attempt to keep most of the code as is and just refactor it into one singular thing, I decided to not potentially break this behavior ...

**Pull request merged**